R/mergedatasetsbycase.R

```diff
@@ -0,0 +1,2238 @@
+        setequal(unique.v.types, c("Date/Time", "Date", "Numeric")) ||
+        setequal(unique.v.types, c("Date/Time", "Text", "Date", "Numeric")))
+        return(.combineVar(AsDateTime))
+    else if (setequal(unique.v.types, c("Date")) ||
+             setequal(unique.v.types, c("Date", "Text")) ||
+             setequal(unique.v.types, c("Date", "Numeric")) ||
+             setequal(unique.v.types, c("Date", "Text", "Numeric")))
+        return(.combineVar(AsDate))
+    else if (setequal(unique.v.types, c("Duration")) ||
+             setequal(unique.v.types, c("Duration", "Text")))
+        return(.combineVar(as.difftime))
+    else if (setequal(unique.v.types, c("Numeric")) ||
+             setequal(unique.v.types, c("Numeric", "Text")))
+    {
+        text.ind <- which(v.types == "Text")
+        is.parsable <- all(vapply(text.ind, function(j) {
+            isParsableAsNumeric(var.list[[j]])
+        }, logical(1)))
+
+        if (is.parsable)
+        {
+            result <- .combineVar(as.numeric)
+            if (isIntegerValued(result))
+                return(as.integer(result))
+            else
+                return(result)
+        }
+        else
+        {
+            return(unlist(lapply(seq_len(n.data.sets), function(i) {
+                v <- var.list[[i]]
+                if (is.null(v))
+                    rep(NA, nrow(data.sets[[i]]))
+                else if (v.types[i] == "Numeric")
+                    as.character(v)
+                else
+                    v
+            })))
+        }
+    }
+    else if (setequal(unique.v.types, c("Text")))
+        return(.combineVar(function(x) x))
+    else if ("Categorical" %in% unique.v.types)
+    {
+        # If there are any categorical variables, convert everything into text.
+        # This only occurs when categorical is combined with date, date/time or
+        # duration variables or there are too many unique values in numeric or
+        # text variables.
+        return(unlist(lapply(seq_len(n.data.sets), function(i) {
+            v <- var.list[[i]]
+            if (is.null(v))
+                rep(NA_character_, nrow(data.sets[[i]]))
+            else if (v.types[i] == "Categorical")
+            {
+                result <- rep(NA_character_, nrow(data.sets[[i]]))
+                val.attr <- attr(var.list[[i]], "labels")
+                for (j in seq_along(val.attr))
+                    result[v == val.attr[j]] <- names(val.attr)[j]
+                result
+            }
+            else
+                as.character(v)
+        })))
+    }
+    else
+    {
+        # Don't expect this to ever occur
+        stop("Unhandled variable types combination: ",
+             paste0(unique.v.types, collapse = ", "))
+    }
+}
+
+# Remap values in variable given a value map
+remapValuesInVariable <- function(variable, map)
+{
+    result <- variable
+    for (i in seqRow(map))
+        result[variable == map[i, 1]] <- map[i, 2]
+    result
+}
```

Comment:
I'm not sure your example will be a problem since the `variable` variable is used to determine the index to modify in `result`, and `variable` never gets modified. It is true that order of mappings may matter if the original index appears in multiple maps (many to one) but this will not be the case with the mappings passed to this function.